sdk/lib/crt/misc/dbgrpt.cpp

```diff
@@ -0,0 +1,358 @@
+    char szCompleteMessage[(DBGRPT_MAX_BUFFER_SIZE+1)*2] = {0}; // The output for debug / file
+
+    if (reportType < 0 || reportType >= _CRT_ERRCNT)
+        return -1;
+
+    if (reportType == _CRT_ASSERT)
+    {
+        if (_InterlockedIncrement(&_CrtInAssert) > 0)
+        {
+            char LineBuffer[20] = {0};
+
+            _itoa(linenumber, LineBuffer, 10);
+
+            OutputDebugStringA("Nested Assert from File: ");
+            OutputDebugStringA(filename ? filename : "<unknown file>");
+            OutputDebugStringA(", Line: ");
+            OutputDebugStringA(LineBuffer);
+            OutputDebugStringA("\n");
+
+            _CrtDbgBreak();
+
+            _InterlockedDecrement(&_CrtInAssert);
+            return -1;
+        }
+    }
+
+    if (filename)
+    {
+        _snprintf(szCompleteMessage, DBGRPT_MAX_BUFFER_SIZE, "%s(%d) : ", filename, linenumber);
+    }
+
+    if (format)
+    {
+        va_list arglist;
+        va_start(arglist, format);
+        int len = _vsnprintf(szFormatted, DBGRPT_MAX_BUFFER_SIZE - 2 - sizeof(DBGRPT_ASSERT_PREFIX_MESSAGE), format, arglist);
+        va_end(arglist);
+
+        if (len < 0)
+        {
+            strcpy(szFormatted, DBGRPT_STRING_TOO_LONG);
+        }
+
+        if (reportType == _CRT_ASSERT)
+            strcat(szCompleteMessage, DBGRPT_ASSERT_PREFIX_MESSAGE);
+        strcat(szCompleteMessage, szFormatted);
+    }
+    else if (reportType == _CRT_ASSERT)
+    {
+        strcat(szCompleteMessage, DBGRPT_ASSERT_PREFIX_NOMESSAGE);
+    }
+
+    if (reportType == _CRT_ASSERT)
+    {
+        if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_FILE)
+            strcat(szCompleteMessage, "\r");
+        strcat(szCompleteMessage, "\n");
+    }
+
+    // FIXME: Handle user report hooks here
+
+
+    if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_FILE)
+    {
+        OutputDebugStringA("ERROR: Please implement _CrtSetReportFile first\n");
+        _CrtDbgBreak();
+    }
+
+    if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_DEBUG)
+    {
+        OutputDebugStringA(szCompleteMessage);
+    }
+
+    int nResult = FALSE;
+    if (_CrtModeOutputFormat[reportType] & _CRTDBG_MODE_WNDW)
+    {
+        nResult = _CrtDbgReportWindow(reportType, filename, linenumber, moduleName, szFormatted);
+
+        switch (nResult)
+        {
+        case IDRETRY:
+            nResult = TRUE;
+            break;
+        case IDIGNORE:
+        default:
+            nResult = FALSE;
+            break;
+        case IDABORT:
+            raise(SIGABRT);
+            _exit(3);
+            break;
+        }
+    }
+
+    if (reportType == _CRT_ASSERT)
+        _InterlockedDecrement(&_CrtInAssert);
+
+    return nResult;
+}
+
+EXTERN_C
+int __cdecl _CrtDbgReportW(int reportType, const wchar_t *filename, int linenumber, const wchar_t *moduleName, const wchar_t *format, ...)
+{
+    wchar_t szFormatted[DBGRPT_MAX_BUFFER_SIZE+1] = {0};           // The user provided message
```

Comment:
Having to duplicate all the code for A/W is a little bit sad.
You could make a templated version and have both functions just call that. Though that might even end up hurting readability, so I don't necessarily have a better suggestion.

Reply:
Initially I wrote this entire thing as a template having the entrypoints simply call that,
however due to the amount of constant strings / helper functions required this became a nightmare.

Reply:
I combined some more code.